advisories/a64325.md

```diff
@@ -0,0 +1,41 @@
+---
+title: Technical Advisory 64325
+summary: Race condition between long running reads and replica removal
+toc: true
+---
+
+Publication date: May 3, 2021
+
+## Description
+
+Cockroach Labs has discovered a race condition, where a long running read request submitted during replica removal may be evaluated on the removed replica, returning an empty result. Common causes of replica removal are range merges and replica rebalancing in the cluster. When a replica is removed there is a small window, where the data is already deleted but a read request will still see it as a valid replica. In this case a read request will always return an empty result for the part of the query that relied on this range.
+
+This is a very rare scenario that can only be encountered, if a number of very unlikely conditions are met:
+
+ - A read request checks that the node is holding the lease and begins processing
+ - The node loses the lease before the read request is able to lock the replica for reading
+ - The read request checks the state of the replica and makes sure it is alive
+ - The new leaseholder decides to immediately removes the old leaseholder replica from the range, due to a rebalance or range merge.
+ - The removal is processed before the read requests starts reading the data from the replica
+ - The read request observes the replica in a post-deletion state returning an empty result set.
+
+ Due to the nature of this race condition, we believe that it is only likely for this to happen on an extremely overloaded node that is struggling to keep up with processing requests and experiences natural delays between each step in the processing of a read request.
+
+This issue affects CockroachDB [versions](/docs/releases/) 19.2 and later.
+
+## Statement
+This is resolved in CockroachDB by PR [#64324], which fix the race condition between read requests and replica removal.
+
+The fix has been applied to maintenance releases of CockroachDB v20.1, v20.2, and v21.1.
+
+This public issue is tracked by [#64325].
+
+## Mitigation
+
+Users of CockroachDB v19.2, v20.1, or v20.2 are invited to upgrade to v20.1.16, v20.2.9, or a later version.
+
+This issue may impact the backups of CockroachDB. You can take the following steps to validate backups to make sure they are correct and complete. TODO: @mattsherman.
```

Comment:
@pbardea Would you fill this in? I assume it’s a summary of this: https://github.com/cockroachlabs/support/issues/940#issuecomment-828814334

It’s important to say what the query can or can’t detect.

--

Format-wise, because these instructions are long, maybe they don’t belong in this document.

Should they be added as a comment on the public issue? https://github.com/cockroachdb/cockroach/issues/64325

Thoughts @lunevalex ?

Reply:
Will follow up. There are a few queries that can help detect these missing from full backups. I've posted them in https://github.com/cockroachlabs/support/issues/940#issuecomment-830420092. They each have tradeoffs in terms of how many edge cases they consider and the number of false positives returned. They're all fairly clunky, especially to explain in general terms in a TA.

Perhaps the best solution here is to find a way to detect these issues with a high-probability but with minimal false-positives reported?

Of note: I don't think this issue is detectable at all in incremental backups.

Reply:
It might be a better decision not to include this in the advisory for now, too much detail. We’ll use the above comment as a resource for the support team, and think about where a public version would make sense.

Reply:
I think we can go with a link and keep the advisory simple.